### PR TITLE
fix(35954): Change spelling for private field incorrectly fixes to a string property

### LINKED ISSUE #35954: Change spelling for private field incorrectly fixes to a string property
**TypeScript Version:**  3.8.0-dev.20191231

**Search Terms:**

private field change spelling, private field autocorrect

**Code**

```ts
class Foo {
    #foobar: number;

    constructor() {
        this.foobar = 12;
    }
}
```

**Actual behavior:**

This code contains a mistake, the suggested option of `Change spelling to '#foobar'` converts the statement into `this["#foobar"]` which is not correct.

**Expected behavior:**

The code should be fixed to `this.#foobar`.

**Playground Link:** 

https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191231&ssl=6&ssc=14&pln=6&pc=20#code/FAYwNghgzlAEBiB7RsDexadgYgGbICMIAnALlgDsBXAWwIFNiBuYDLERCqAF2KpG6JiACgCUaNlizcAFgEsoAOnyIixWAF5YARgBMLKQF9ghoA

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #35954

### CODE REVIEW

File: src/services/codefixes/fixSpelling.ts [hunk 1]
```diff
@@ -62,7 +62,11 @@ namespace ts.codefix {
     }
 
     function doChange(changes: textChanges.ChangeTracker, sourceFile: SourceFile, node: Node, suggestion: string, target: ScriptTarget) {
-        if (!isIdentifierText(suggestion, target) && isPropertyAccessExpression(node.parent)) {
+        if (
+            !isIdentifierText(suggestion, target) &&
+            !isPrivatePropertyAccessLike(suggestion) &&
+            isPropertyAccessExpression(node.parent)
```
Comment (ajafff) [inline src/services/codefixes/fixSpelling.ts]: Consider supporting ElementAccessExpression, too

File: tests/cases/fourslash/codeFixSpellingPrivatePropertyName.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/// <reference path='fourslash.ts' />
+
+////class A {
+////   #foo: number;
```
Comment (ajafff) [inline tests/cases/fourslash/codeFixSpellingPrivatePropertyName.ts]: What happens if there is a property named `"#foo"` instead?

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
This case seems to cause a crash

```ts
class C {
    "#foo" = 10;
    blah() {
        this.foo/*requestsQuickfixesHere*/
    }
}
```

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser I see that for this case QFix works incorrectly 

```TS
class C {
    "#foo" = 10;
    blah() {
        this.#foo
    }
}
```

I did not find the existing helper to check is the suggestion `string` is a private field. Maybe you can suggest to me the way do it?

File: tests/cases/fourslash/codeFixSpellingPrivatePropertyName.ts (hunk 2 above)
Comment (a-tarasyuk) [reply tests/cases/fourslash/codeFixSpellingPrivatePropertyName.ts]: @ajafff Good point, I didn't test that case.

**Comment by DanielRosenwasser** [comment]
My mistake - it seems to be separate from this, and only reproduces on the playground.

I'm not totally sure though. I would guess that you need to find a way to figure it out by walking up the AST and trying to resolve yourself - alternatively, the `getSuggestion` functions would need to return a TypeScript `Symbol` instead of a `string`.

**Comment by DanielRosenwasser** [comment]
Maybe @sandersn can help here.

File: src/services/codefixes/fixSpelling.ts [hunk 3]
```diff
@@ -92,4 +96,8 @@ namespace ts.codefix {
 
         return context.program.getSourceFile(resolvedModule.resolvedFileName);
     }
+
+    function isPrivatePropertyAccessLike(name: string): boolean {
+        return name.charCodeAt(0) === CharacterCodes.hash && name.length > 1;
```
Comment (sandersn) [inline src/services/codefixes/fixSpelling.ts]: You can either 
1. Look at getSuggestionForNonExistentSymbol and make a version that distinguishes between private identifiers and normal ones in some way. That's probably the best way, and the infrastructure is probably mostly in place. I'll take a look at how to do that when I get to the office.
2. `getPropertyOfType(getTypeOfSymbol(getSymbolOfNode((node.parent as PropertyAccessExpression).expression)), name)` and then check whether the property is private. I can't remember the best way to do that; you might check whether .valueDeclaration.kind is SyntaxKind.PrivateIdentifier.

Comment (a-tarasyuk) [reply src/services/codefixes/fixSpelling.ts]: @sandersn Thanks. I checked the `getPropertyOfType` and seems it doesn't handle private identifiers., I found `getPrivateIdentifierPropertyOfType` which I applied instead.

**Comment by DanielRosenwasser** [comment]
Does any of our external code depend on the old `string` suggestion APIs?

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser I think that these three methods  (`getSuggestionForNonexistentProperty`, `getSuggestionForNonexistentExport`, `getSuggestionForNonexistentSymbol`) are used only in the fixSpelling service and internally in checker.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by sandersn** [comment]
@a-tarasyuk ping me when you've merged from master and I'll merge this.

**Comment by a-tarasyuk** [comment]
@sandersn I've merged the latest changes from the `master` ;)

**Pull request merged**